### DISCUSSION TITLE
post-request script

### DISCUSSION THREAD

**Discussion by petoc** [Kong/insomnia#6242] [Ideas] [answered]
It would be great if Insomnia supported such thing. It is available in Postman (tab Tests) and allows for example setting environment variable from response. Insomnia support single environment variable from single requests, but sometimes it is required to set same environment variable from multiple requests (i.e. multiple variations of login requests that could set access token to environment variable, so it could be used by other requests). There is also issue requesting such/similar feature (#1968). It is also similar to discussed pre-request scripts in #3806.

There are probably many ways how to implement this, but personaly I find scripts to be most transparent, because all settings are within request configuration and are transfered upon reqest duplication and so on.
Maybe there could be even one tab Scripts and it could hold both pre-request and post-request scripts.

Are there any plans for such feature?

You can find my attempt on post request script at https://github.com/petoc/insomnia/tree/feature-script .
It definitely needs code refining.
Node's `vm` is used to execute javascript and exposes very basic insomnia API and `console`. Script is saved for individual requests and should be also exported/imported together with requests.
Probably `isolated-vm` would be better and more secure alternative to node's `vm`, but I had some trouble getting it to work.

![insomnia-script](https://github.com/Kong/insomnia/assets/63091190/a89a438f-2bda-499e-ac29-6a37148f3f74)
![insomnia-script2](https://github.com/Kong/insomnia/assets/63091190/c4516529-7f1d-4649-803e-15498aacc5c5)

**Comment by deepanshuds** [comment]
Hello @subnetmarco, 

are there any plans to include this post-request script feature in the upcoming releases?

**Reply by subnetmarco** [reply]
Yes: we are shipping pre-request scripting in 9.0 in just a couple of weeks (a beta is available now), and we will start work on post-request scripting right after.

**Reply by wwc7654321** [reply]
I thought it would be available the same time with pre-request script. Looking forward to next version

**Reply by subnetmarco** [reply]
It will ship very shortly afterwards since the underlying engine is the same.

**Comment by subnetmarco** [comment]
To recap in the main thread, this capability will ship in a minor 9.1 immediately after 9.0 (when we ship pre-request scripting).

**Reply by subnetmarco** [reply]
Since 9.0 has been released today, work on post-scripts will start now and we expect it to be release in approximately a month 🤞

**Comment by wzcskhc** [comment]
![Screen-Record-2024-04-25 at 13 20 20@2x](https://github.com/Kong/insomnia/assets/5238990/a8830b24-4db6-430f-b24e-5b6081452c95)
`require('crypto-js')` is  not work

**Reply by ihexxa** [reply]
Started #7315 for this, thanks for the feedback.

**Comment by subnetmarco** [comment]
This issue has been fixed in 9.1

**Reply by petoc** [reply]
But post request script feature is still missing, right?

**Reply by subnetmarco** [reply]
@petoc post request scripting be available in a beta in around a couple of weeks.

**Reply by alaz-aura** [reply]
looking forward to this!

**Comment by mumi** [comment]
@subnetmarco @ihexxa Will it be possible to manipulate the response (body/header)? Currently this is not possible current feature branches. Would be interesting, for example, if you want to chain requests to fetch auth tokens where you have to transform the responses

**Reply by ihexxa** [reply]
Hello @mumi , the major part of the post-request script hasn't been merged into the feature branch yet, it will be ready soon, then you'll be able to read `insomnia.response` in it.
Currently you should be able to `sendRequest` and read its response in the pre-request script for the use case you mentioned.

**Reply by mumi** [reply]
Hey @ihexxa, for example for a case like this, where I get a token from an echo endpoint, so unstructured data, where I just want to extract the token and return it as the response body, so that I can set it as a reference (Response raw body) that will always trigger, if I use that env var. The easiest way to do this would be to overwrite the response body in the post-request script

<img width="934" alt="Pasted Graphic" src="https://github.com/Kong/insomnia/assets/17066103/b19b6365-f666-4665-be45-3082abbaa467">

**Reply by ihexxa** [reply]
Hi @mumi, yeah it would be interesting, we will try to deliver it by following existing behavior at first and see if it would be cool to enhance it, as it may create a new way to transfer data between entities.

**Comment by subnetmarco** [comment, marked as answer]
The latest [9.3.0 beta](https://github.com/Kong/insomnia/releases/tag/core%409.3.0-beta.0) includes this capability, please try it out and let us know your feedback!

**Reply by petoc** [reply]
I tried just basic functionality, setting env. variable and it is working as expected.
Thank you, especially @ihexxa for implementing this.

**Reply by subnetmarco** [reply]
Great - This is the time to send more feedback before we wrap it up for 9.3 GA in about 2-3 weeks.

**Reply by carlosaya** [reply]
did you mean to link to https://github.com/Kong/insomnia/releases/tag/core%409.3.0-beta.0 instead?

**Reply by subnetmarco** [reply]
Yes! Updated.

**Comment by 0erty** [comment]
This is useless since setting environment variables does not work via the CLI :(

**Reply by alaz-aura** [reply]
why does that affect post-request usefulness?

**Reply by 0erty** [reply]
I think the main advantage of the post-request function is its use in tests. But now it doesn't work in CLI mode. But that's just my opinion.

**Reply by subnetmarco** [reply]
We are going to support collection runs in the CLI and in the application. In 9.3 we will unblock the capability by introducing after-response scripting, then in 9.4 and above you will see both an in-app collection runner and CLI support to run collections.

**Reply by 0erty** [reply]
I'm waiting for this feature, thanks

**Comment by subnetmarco** [comment]
After response scripting [has been released today](https://konghq.com/blog/product-releases/insomnia-9-3-ga) in Insomnia 9.3 GA.

**Comment by eliseudev** [comment]
> It would be great if Insomnia supported such thing. It is available in Postman (tab Tests) and allows for example setting environment variable from response. Insomnia support single environment variable from single requests, but sometimes it is required to set same environment variable from multiple requests (i.e. multiple variations of login requests that could set access token to environment variable, so it could be used by other requests). There is also issue requesting such/similar feature (#1968). It is also similar to discussed pre-request scripts in #3806.
> 
> There are probably many ways how to implement this, but personaly I find scripts to be most transparent, because all settings are within request configuration and are transfered upon reqest duplication and so on. Maybe there could be even one tab Scripts and it could hold both pre-request and post-request scripts.
> 
> Are there any plans for such feature?
> 
> You can find my attempt on post request script at https://github.com/petoc/insomnia/tree/feature-script . It definitely needs code refining. Node's `vm` is used to execute javascript and exposes very basic insomnia API and `console`. Script is saved for individual requests and should be also exported/imported together with requests. Probably `isolated-vm` would be better and more secure alternative to node's `vm`, but I had some trouble getting it to work.
> 
> ![insomnia-script](https://private-user-images.githubusercontent.com/63091190/259312688-a89a438f-2bda-499e-ac29-6a37148f3f74.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk5MzYwMzAsIm5iZiI6MTcxOTkzNTczMCwicGF0aCI6Ii82MzA5MTE5MC8yNTkzMTI2ODgtYTg5YTQzOGYtMmJkYS00OTllLWFjMjktNmEzNzE0OGYzZjc0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAyVDE1NTUzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQzMzVhMmNlYjkyYTY4ZTA3YzcxYjMwYTU1OGEwZjZlNjkwMmYxOGM1NDIyMjFmZTZmMThmYTlkMTkyY2ZjYzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kYHunrKH78xjCMF77sIcQOOWovdv12kBuwi0OeMdXEY) ![insomnia-script2](https://private-user-images.githubusercontent.com/63091190/259312805-c4516529-7f1d-4649-803e-15498aacc5c5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk5MzYwMzAsIm5iZiI6MTcxOTkzNTczMCwicGF0aCI6Ii82MzA5MTE5MC8yNTkzMTI4MDUtYzQ1MTY1MjktN2YxZC00NjQ5LTgwM2UtMTU0OThhYWNjNWM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAyVDE1NTUzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkYTQzNjZkOTMxNjBkNDliMTE0ZmQ1Nzk1YTczYmM0YTM3NzQxZDQwZjMzYzI0ZDM5ZmQ0Yzc5NzllZDg1ZDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.dOtxBjprbLpvAIj3eXMYJ5bo16PxJlszZBswwxfs2HY)


var data = insomnia.response.json();
if(data && data.jwt){
	insomnia.environment.set('token', data.jwt);
}
